File: app/src/main/java/com/example/javatokotlin/retrofit/RetrofitClient.java
package com.example.javatokotlin.retrofit;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;


public class RetrofitClient {

    private static final String BASE_URL = "https://api.github.com/";
    private static Retrofit retrofit = null;

    public static GithubAPIService getGithubAPIService() {
        return getClient().create(GithubAPIService.class);
    }

    private static Retrofit getClient() {
        if (retrofit == null) {
            retrofit = new Retrofit.Builder()
                    .baseUrl(RetrofitClient.BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        }
        return retrofit;
    }
}

File: app/src/main/java/com/example/javatokotlin/retrofit/RetrofitClient.kt
package com.example.javatokotlin.retrofit

import retrofit2.Retrofit
import com.example.javatokotlin.retrofit.GithubAPIService
import com.example.javatokotlin.retrofit.RetrofitClient
import retrofit2.converter.gson.GsonConverterFactory

//Object declarations are initialized lazily when accessed
object RetrofitClient {
    private const val BASE_URL = "https://api.github.com/"

    // object declaration is initialized lazy by default
    private val retrofit: Retrofit =
        Retrofit.Builder()
            .baseUrl(BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()

    val githubAPIService: GithubAPIService by lazy {
        retrofit.create(GithubAPIService::class.java)

    }

}